src/lib/core/fiber.c

```diff
@@ -620,8 +628,25 @@ fiber_join(struct fiber *fiber)
 			 * to put it back in.
 			 */
 			rlist_add_tail_entry(&fiber->wake, fiber(), state);
-			fiber_yield();
-		} while (! fiber_is_dead(fiber));
+			if (timeout != TIMEOUT_INFINITY) {
+				double time = fiber_time();
+				exceeded = fiber_yield_timeout(timeout);
+				timeout -= (fiber_time() - time);
+			} else {
+				fiber_yield();
+			}
+		} while (! fiber_is_dead(fiber) && ! exceeded && timeout > 0);
+	}
+
+	if (! fiber_is_dead(fiber)) {
+#if defined (TARGET_OS_FREEBSD) || defined (TARGET_OS_NETBSD) || \
```

Comment:
Why do you set errno? You don't check it and set diag message anyway.

Reply:
We can only been here if we didn't wait for the fiber to finish, so i set errno